python/ray/actor_mesh.py

```diff
@@ -0,0 +1,189 @@
+        self.actor_mesh = actor_mesh
+
+    def __getattr__(self, k):
+        return ActorMethodProxy(self.actor_mesh, k)
+    
+class ActorMethod(Generic[P, R]):
+
+    def __init__(self, method):
+        self.method = method
+
+    def all(self, *args: P.args, **kwargs: P.kwargs) -> List["ray.ObjectRef[R]"]:
+        "Call all the actors in the mesh with the same arguments."
+        pass
+
+    def choose(self, *args: P.args, **kwargs: P.kwargs) -> "ray.ObjectRef[R]":
+        "Call one of the actors in the mesh."
+        pass
+        
+    def shard(self, *args: P.args, **kwargs: P.kwargs) -> R:
+        "Call all the actors in the mesh, shard the argument(s) among them."
+        pass
+    
+class ActorMethodProxy:
+
+    def __init__(self, actor_mesh: "ActorMesh", actor_method: str) -> None:
+        self.actor_mesh = actor_mesh
+        self.actor_method = actor_method
+
+    def all(self, *args, **kwargs):
+        results = []
+        for actor in self.actor_mesh.actors:
+            method = getattr(actor, self.actor_method)
+            results.append(method.remote(*args, **kwargs))
+        return results
+    
+    def choose(self, *args, **kwargs):
+        # Choosing randomly right now, this is bad and needs to be improved
+        actor = random.choice(self.actor_mesh.actors)
+        method = getattr(actor, self.actor_method)
+        return method.remote(*args, **kwargs)
+    
+    def shard(self, *args, **kwargs):
+        orig_func = getattr(self.actor_mesh._actor_cls, self.actor_method)
+        dispatch_fn = getattr(orig_func, "_dispatch_fn", None)
+        assert dispatch_fn, "Sharding only supported if dispatch_fn is specified (later we will have a good default one)"
+        return dispatch_fn(self.actor_mesh, self.actor_method, *args, **kwargs)
+
+
+class ActorMesh(Generic[T]):
+
+    def __init__(
+            self,
+            actor_cls: Callable[..., T],
+            args: Any,
+            kwargs: Any,
+            shape: Union[int, Tuple[int], Dict[str, int]],
+            # TODO: Not used yet
+            resources_per_actor: Optional[Dict[str, float]] = None,
+            runtime_env: Optional[Dict[str, Any]] = None,
+        ):
+
+        if isinstance(shape, int):
+            shape = (shape,)
+
+        self.shape = shape
+
+        if isinstance(shape, dict):
+            self._num_actors = sum(shape.values())
+        else:
+            self._num_actors = sum(shape)
+
+        self._actor_cls = actor_cls
+        self._actors = []
+        runtime_env = copy.deepcopy(runtime_env) or {}
+        for i in range(self._num_actors):
+            env_vars = {**runtime_env.get("env_vars", {}), "RAY_ACTOR_MESH_RANK": str(i)}
+            ray_actor_cls = ray.remote(
+                runtime_env={**runtime_env, "env_vars": env_vars},
+            )(actor_cls)
+            actor = ray_actor_cls.remote(*args, **kwargs)
+            self._actors.append(actor)
+    
+    @property
+    def methods(self) -> type[T]:
```

Comment:
This is currently introduced to (1) provide correct typing for the methods (the original idea for this was from https://github.com/zen-xu/sunray, see also https://github.com/ray-project/ray/issues/54149) and (2) avoid clashes of user defined methods and methods / properties defined on the `ActorMesh` class itself.

However it is slightly clunky having to always type `mesh.methods.{method}`, so ideally I'd like to get rid of this property and provide the actor methods directly as methods on the `ActorMesh[ActorClass]` class. (2) could be solved in other ways.

Reply:
> However it is slightly clunky having to always type mesh.methods.{method}, so ideally I'd like to get rid of this property and provide the actor methods directly as methods on the ActorMesh[ActorClass] class.

+1